kernel/src/table_changes/log_replay.rs

```diff
@@ -52,251 +55,258 @@ pub(crate) fn table_changes_action_iter(
+    let action_iter = engine.get_json_handler().read_json_files(
+        &[commit_file.location.clone()],
+        visitor_schema,
+        None, // not safe to apply data skipping yet
+    )?;
 
-        let mut remove_dvs = HashMap::default();
-        let mut add_paths = HashSet::default();
-        let mut has_cdc_action = false;
-        for actions in action_iter {
-            let actions = actions?;
+    let mut remove_dvs = HashMap::default();
+    let mut add_paths = HashSet::default();
+    let mut has_cdc_action = false;
+    for actions in action_iter {
+        let actions = actions?;
 
-            let mut visitor = PreparePhaseVisitor {
-                add_paths: &mut add_paths,
-                remove_dvs: &mut remove_dvs,
-                has_cdc_action: &mut has_cdc_action,
-                protocol: None,
-                metadata_info: None,
-            };
-            visitor.visit_rows_of(actions.as_ref())?;
+        let mut visitor = ProcessCdfCommitVisitor {
+            add_paths: &mut add_paths,
+            remove_dvs: &mut remove_dvs,
+            has_cdc_action: &mut has_cdc_action,
+            protocol: None,
+            metadata: None,
+        };
 
-            if let Some(protocol) = visitor.protocol {
-                ensure_cdf_read_supported(&protocol)
-                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
-            }
-            if let Some((schema, configuration)) = visitor.metadata_info {
-                let schema: StructType = serde_json::from_str(&schema)?;
-                // Currently, schema compatibility is defined as having equal schema types. In the
-                // future, more permisive schema evolution will be supported.
-                // See: https://github.com/delta-io/delta-kernel-rs/issues/523
-                require!(
-                    table_schema.as_ref() == &schema,
-                    Error::change_data_feed_incompatible_schema(table_schema, &schema)
-                );
-                let table_properties = TableProperties::from(configuration);
-                check_cdf_table_properties(&table_properties)
-                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
+        visitor.visit_rows_of(actions.as_ref())?;
+        let metadata_changed = visitor.metadata.is_some();
+        match (visitor.protocol, visitor.metadata) {
+            (None, None) => {} // no change
+            (protocol, metadata) => {
+                // at least one of protocol and metadata changed, so update the table configuration
+                *table_configuration = TableConfiguration::try_new(
+                    metadata.unwrap_or_else(|| table_configuration.metadata().clone()),
+                    protocol.unwrap_or_else(|| table_configuration.protocol().clone()),
+                    table_configuration.table_root().clone(),
+                    commit_file.version,
+                )?;
```

Comment:
related to comment above - this feels like the need for an API that will allow us to do better here instead of cloning